plugins/xml/core/src/test/java/org/pentaho/di/trans/steps/xmljoin/XmlJoinMetaGetFieldsTest.java

```diff
@@ -74,4 +74,65 @@ public void testGetFieldsReturnTargetStepFieldsPlusResultXmlField() throws Excep
     Assert.assertEquals( targetStepField, strings[0] );
     Assert.assertEquals( resultXmlFieldName, strings[1] );
   }
+
+  @Test
+  public void testGetFieldsReturnTargetStepFieldsWithDuplicates() throws Exception {
+    // Source Step
+    String sourceXmlStep = "source xml step name";
+    String sourceStepField1 = "a";
+    String sourceStepField2 = "b";
+
+    // Target Step
+    String targetXmlStep = "target xml step name";
+    String targetStepField1 = "b";
+    String targetStepField2 = "c";
+
+    // XML Join Result
+    String resultXmlFieldName = "result xml field name";
+
+    // Source Row Meta
+    RowMeta rowMetaPreviousSourceStep = new RowMeta();
+    rowMetaPreviousSourceStep.addValueMeta( new ValueMeta( sourceStepField1, ValueMetaInterface.TYPE_STRING ) );
+    rowMetaPreviousSourceStep.addValueMeta( new ValueMeta( sourceStepField2, ValueMetaInterface.TYPE_STRING ) );
+
+    // Set source step in XML Join step.
+    xmlJoinMeta.setSourceXMLstep( sourceXmlStep );
+    StepMeta sourceStepMeta = new StepMeta();
+    sourceStepMeta.setName( sourceXmlStep );
+
+    when( transMeta.findStep( sourceXmlStep ) ).thenReturn( sourceStepMeta );
+    when( transMeta.getStepFields( sourceStepMeta, null, null ) ).thenReturn( rowMetaPreviousSourceStep );
+
+    // Target Row Meta
+    RowMeta rowMetaPreviousTargetStep = new RowMeta();
+    rowMetaPreviousTargetStep.addValueMeta( new ValueMeta( targetStepField1, ValueMetaInterface.TYPE_STRING ) );
+    rowMetaPreviousTargetStep.addValueMeta( new ValueMeta( targetStepField2, ValueMetaInterface.TYPE_STRING ) );
+
+    // Set target step in XML Join step.
+    xmlJoinMeta.setTargetXMLstep( targetXmlStep );
+    StepMeta targetStepMeta = new StepMeta();
+    targetStepMeta.setName( targetXmlStep );
+
+    when( transMeta.findStep( targetXmlStep ) ).thenReturn( targetStepMeta );
+    when( transMeta.getStepFields( targetStepMeta, null, null ) ).thenReturn( rowMetaPreviousTargetStep );
+
+    // Set result field name
+    xmlJoinMeta.setValueXMLfield( resultXmlFieldName );
+
+    RowMeta rowMeta = new RowMeta();
+    ValueMetaString removeValueMeta1 = new ValueMetaString( "a" );
+    rowMeta.addValueMeta( removeValueMeta1 );
+    ValueMetaString keepValueMeta1 = new ValueMetaString( "b" );
+    rowMeta.addValueMeta( keepValueMeta1 );
+    ValueMetaString keepValueMeta2 = new ValueMetaString( "c" );
+    rowMeta.addValueMeta( keepValueMeta2 );
+
+    // Get output fields
+    xmlJoinMeta.getFields( rowMeta, "testStepName", null, null, transMeta, null, null );
+    Assert.assertEquals( 3, rowMeta.size() );
+    String[] strings = rowMeta.getFieldNames();
+    Assert.assertEquals( "b", strings[0] );
```

Comment:
Please use an `import static` for these like you used for the `when` methods.
Also, prefer to use `doReturn( something ).when( object ).method()` instead as this syntax has given us some problems in the past with tests failing for no apparent reason.

Reply:
Hi @pentaho-lmartins! I've made the changes you requested. I also updated the above test, so that everything looks consistent.